File: common/src/main/kotlin/net/impleri/playerskills/skills/tiered/TieredSkill.kt
package net.impleri.playerskills.skills.tiered

import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.TeamMode
import net.minecraft.resources.ResourceLocation

open class TieredSkill(
  name: ResourceLocation,
  value: String?,
  description: String?,
  options: List<String>?,
  changesAllowed: Int?,
  teamMode: TeamMode?,
  notify: Boolean?,
  notifyKey: String?,
) : Skill<String>(
  name,
  TieredSkillType.NAME,
  value,
  description,
  options ?: ArrayList(),
  changesAllowed ?: Skill.UNLIMITED_CHANGES,
  teamMode ?: TeamMode.off(),
  notify ?: false,
  notifyKey,
) {
  override fun copy(value: String?, changesAllowed: Int): Skill<String> {
    return TieredSkill(name, value, description, options, changesAllowed, teamMode, notify, notifyKey)
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/tiered/TieredSkill.scala
package net.impleri.playerskills.skills.tiered

import net.impleri.playerskills.api.skills.ChangeableSkillOps
import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.TeamMode
import net.impleri.playerskills.facades.minecraft.core.ResourceLocation

case class TieredSkill(
  override val name: ResourceLocation,
  override val value: Option[String] = None,
  override val description: Option[String] = None,
  override val options: List[String] = List.empty,
  override val changesAllowed: Int = Skill.UNLIMITED_CHANGES,
  override val teamMode: TeamMode = TeamMode.Off(),
  override val announceChange: Boolean = false,
  override val notifyKey: Option[String] = None,
) extends Skill[String] with ChangeableSkillOps[String, TieredSkill] {
  override val skillType: ResourceLocation = TieredSkillType.NAME

  override protected[playerskills] def mutate(value: Option[String], changesAllowed: Int): TieredSkill = {
    copy(
      value = value, changesAllowed = changesAllowed,
    )
  }
}
